code/modules/GOAP/GOAPAgent.dm

```diff
@@ -0,0 +1,228 @@
+			act_state()
+
+/datum/goap_agent/proc/act_state()
+	if(!LAZYLEN(action_queue))
+		brain_state = STATE_IDLE
+		return
+	var/datum/goap_action/curr_action = action_queue[action_queue.len]
+	curr_action = action_queue[action_queue.len]
+	var/range_check = curr_action.IsInRange(agent)
+	if(curr_action.CheckDone(agent))
+		action_queue.len--
+	if(already_acting)
+		return
+	if(!LAZYLEN(action_queue))
+		return
+	if(!range_check)
+		brain_state = STATE_MOVINGTO
+		return
+	already_acting = TRUE
+	if(!curr_action.Perform(agent))
+		goap_debug("PERFORM FAILED [curr_action]")
+		brain_state = STATE_IDLE
+		path = list()
+		info.PlanAborted(curr_action)
+		already_acting = FALSE
+	else
+		goap_debug("PERFORMED [curr_action]")
+		if(action_queue.len == 1 && action_queue[1] == curr_action)
+			brain_state = STATE_IDLE
+			path = list()
+			already_acting = FALSE
+			return
+		path = list()
+		already_acting = FALSE
+
+/datum/goap_agent/proc/moving_state()
+	if(fuck_you_astar)
+		return
+	var/datum/goap_action/curr_action = action_queue[action_queue.len]
+
+	if(curr_action.RequiresInRange(agent) && !curr_action.target)
+		goap_debug("An action ([curr_action]) requires a target, but did not get one set")
+		brain_state = STATE_IDLE
+		return
+	var/dense_garbage = null
+	for(var/obj/I in get_turf(curr_action.target))
+		if(I.density)
+			dense_garbage = 1
+			break
+	var/proc_to_use = /turf/proc/reachableTurftest
+	switch(movement_type)
+		if(1, 4) // AStar, Full
+			if(!path || !path.len)
+				fuck_you_astar = TRUE
+				if(!curr_action.path_to_use)
+					if(!isturf(curr_action.target))
+						path = get_path_to(agent, get_turf(curr_action.target), /turf/proc/Distance_cardinal, 0, 200, adjacent = proc_to_use, id=given_pathfind_access, mintargetdist = dense_garbage)
+					else
+						path = get_path_to(agent, curr_action.target, /turf/proc/Distance_cardinal, 0, 200, adjacent = proc_to_use, id=given_pathfind_access, mintargetdist = dense_garbage)
+					if(!path || !path.len) // still can't path
+						goap_debug("Can't path to plan, giving up")
+						brain_state = STATE_IDLE
+						fuck_you_astar = FALSE
+						return 0
+				else
+					path = curr_action.path_to_use
+			fuck_you_astar = FALSE
+			last_node = get_turf(path[path.len]) //This is the turf at the end of the path, it should be equal to dest.
+			current_loc = get_turf(agent)
+			curr_action.PerformWhileMoving(agent)
+			MoveTo_AStar(curr_action, path)
+		if(2) // AStar, Fake
+			curr_action.PerformWhileMoving(agent)
+			MoveTo_FakeStar(curr_action)
+		if(3) // No Pathfinding, Straight Line
+			curr_action.PerformWhileMoving(agent)
+			MoveTo(curr_action)
+
+/datum/goap_agent/proc/idle_state()
+	var/list/worldstate = info.GetWorldState(src)
+	var/list/goal = info.GetGoal(src)
+
+	var/list/plan = planner.Plan(agent, our_actions, worldstate, goal)
+
+	if(!LAZYLEN(plan))
+		info.PlanFailed(goal)
+		return
+	goap_debug("I am gonna act")
+	goap_debug(plan.len)
+
+	for(var/i in plan)
+		goap_debug(i)
```

Comment:
This is a useless for loop if goap_debug isn't enabled

Reply:
Cant do much about it. Sorry.